src/gbm/gblinear.cc

```diff
@@ -9,188 +9,98 @@
 #include <sstream>
-#include <cstring>
 #include <algorithm>
+#include "../common/timer.h"
 
 namespace xgboost {
 namespace gbm {
 
 DMLC_REGISTRY_FILE_TAG(gblinear);
 
-// model parameter
-struct GBLinearModelParam :public dmlc::Parameter<GBLinearModelParam> {
-  // number of feature dimension
-  unsigned num_feature;
-  // number of output group
-  int num_output_group;
-  // reserved field
-  int reserved[32];
-  // constructor
-  GBLinearModelParam() {
-    std::memset(this, 0, sizeof(GBLinearModelParam));
-  }
-  DMLC_DECLARE_PARAMETER(GBLinearModelParam) {
-    DMLC_DECLARE_FIELD(num_feature).set_lower_bound(0)
-        .describe("Number of features used in classification.");
-    DMLC_DECLARE_FIELD(num_output_group).set_lower_bound(1).set_default(1)
-        .describe("Number of output groups in the setting.");
-  }
-};
-
 // training parameter
 struct GBLinearTrainParam : public dmlc::Parameter<GBLinearTrainParam> {
   /*! \brief learning_rate */
-  float learning_rate;
-  /*! \brief regularization weight for L2 norm */
-  float reg_lambda;
-  /*! \brief regularization weight for L1 norm */
-  float reg_alpha;
-  /*! \brief regularization weight for L2 norm in bias */
-  float reg_lambda_bias;
+  std::string linear_updater;
+  // flag to print out detailed breakdown of runtime
+  int debug_verbose;
+  float tolerance;
   // declare parameters
   DMLC_DECLARE_PARAMETER(GBLinearTrainParam) {
-    DMLC_DECLARE_FIELD(learning_rate).set_lower_bound(0.0f).set_default(1.0f)
-        .describe("Learning rate of each update.");
-    DMLC_DECLARE_FIELD(reg_lambda).set_lower_bound(0.0f).set_default(0.0f)
-        .describe("L2 regularization on weights.");
-    DMLC_DECLARE_FIELD(reg_alpha).set_lower_bound(0.0f).set_default(0.0f)
-        .describe("L1 regularization on weights.");
-    DMLC_DECLARE_FIELD(reg_lambda_bias).set_lower_bound(0.0f).set_default(0.0f)
-        .describe("L2 regularization on bias.");
-    // alias of parameters
-    DMLC_DECLARE_ALIAS(learning_rate, eta);
-    DMLC_DECLARE_ALIAS(reg_lambda, lambda);
-    DMLC_DECLARE_ALIAS(reg_alpha, alpha);
-    DMLC_DECLARE_ALIAS(reg_lambda_bias, lambda_bias);
-  }
-  // given original weight calculate delta
-  inline double CalcDelta(double sum_grad, double sum_hess, double w) const {
-    if (sum_hess < 1e-5f) return 0.0f;
-    double tmp = w - (sum_grad + reg_lambda * w) / (sum_hess + reg_lambda);
-    if (tmp >=0) {
-      return std::max(-(sum_grad + reg_lambda * w + reg_alpha) / (sum_hess + reg_lambda), -w);
-    } else {
-      return std::min(-(sum_grad + reg_lambda * w - reg_alpha) / (sum_hess + reg_lambda), -w);
-    }
-  }
-  // given original weight calculate delta bias
-  inline double CalcDeltaBias(double sum_grad, double sum_hess, double w) const {
-    return - (sum_grad + reg_lambda_bias * w) / (sum_hess + reg_lambda_bias);
+    DMLC_DECLARE_FIELD(linear_updater)
+        .set_default("updater_shotgun")
+        .describe("Update algorithm for linear model.");
```

Comment:
It would be very helpful when descriptions would include a list of current available options.
Another approach that would keep things clear is to use enum.